Aj/entitlement session backend

staubina opened the pull request:
This PR creates the back end for enroll, unenroll, and switch session for the Course Dashboard.

LEARNER-3136, and LEARNER-2668
https://openedx.atlassian.net/browse/LEARNER-3136
https://openedx.atlassian.net/browse/LEARNER-2669

staubina commented:
jenkins run all

staubina commented on common/djangoapps/enrollment/views.py:
```diff
@@ -612,60 +633,87 @@ def post(self, request):
                 }
                 consent_client.provide_consent(**kwargs)
 
-            enrollment_attributes = request.data.get('enrollment_attributes')
```
This section was not removed.  I changed the indentation and shifted it down a bit.

staubina commented on common/djangoapps/enrollment/views.py:
```diff
@@ -612,60 +633,87 @@ def post(self, request):
+            # Add Enrollment for the Entitlement user with the correct Mode
+            # This should only occur if the User has a Course Entitlement in place.
+            # As a reault the api_key_permissions do not apply the User may enroll themselves based on the entitlement.
+            if course_entitlement and is_active:
```
Note: This is the enroll case for entitlements.

staubina commented on common/djangoapps/enrollment/views.py:
```diff
@@ -612,60 +633,87 @@ def post(self, request):
-            else:
-                # Will reactivate inactive enrollments.
-                response = api.add_enrollment(
+                CourseEntitlement.set_enrollment(
```
Note: This is the unenroll case for entitlements.

staubina commented on common/djangoapps/enrollment/views.py:
```diff
@@ -612,60 +633,87 @@ def post(self, request):
+                    enrollment=CourseEnrollment.get_enrollment(user, course_id)
+                )
+                log.info('Enrolling [%s] entitlement for run [%s] of Course [%s].', username, course_id, course_uuid)
+            elif course_entitlement and is_active is not None and not is_active:
```
Note: This is the unenroll case for entitlements.

staubina commented on common/djangoapps/enrollment/views.py:
```diff
@@ -612,60 +633,87 @@ def post(self, request):
                 )
+                CourseEntitlement.set_enrollment(course_entitlement, None)
+                log.info('Unenrolling [%s] entitlement for run [%s] of Course [%s].', username, course_id, course_uuid)
+            else:
```
Note: The existing enroll logic that still requires staff level access.

staubina commented:
jenkins run a11y

staubina commented:
jenkins run quality

staubina commented:
@bderusha @mjfrey Please take a look at this PR when you have a moment.

I am working on Tests now.  I just want to get the new API looked at sooner than later.  Also, @HarryRein Is looking at integration today.

FYI @edx/helio

mjfrey commented on common/djangoapps/entitlements/models.py:
```diff
@@ -11,7 +11,7 @@ class CourseEntitlement(TimeStampedModel):
 
     user = models.ForeignKey(settings.AUTH_USER_MODEL)
-    uuid = models.UUIDField(default=uuid_tools.uuid4, editable=False)
+    uuid = models.UUIDField(default=uuid_tools.uuid4, editable=False, unique=True)
```
do you need a migration for this?

staubina replied:
It should be in this PR.

staubina replied:
The file, common/djangoapps/entitlements/migrations/0003_auto_20171120_1432.py is included in the PR with the migration.

bderusha commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+            )
+
+        if entitlement.enrollment_course_run is None:
+            return Response()
```
Any reason to not have this also be 204?

bderusha commented on common/djangoapps/entitlements/models.py:
```diff
@@ -24,3 +24,25 @@ class CourseEntitlement(TimeStampedModel):
+    @classmethod
+    def get_active_user_course_entitlements(cls, user, course_uuid):
+        """
+        Returns all the available sessions for a given course.
```
Update docstring

staubina replied:
Dead code that is not deleted 🗑

amangano-edx commented on common/djangoapps/entitlements/models.py:
```diff
@@ -24,3 +24,25 @@ class CourseEntitlement(TimeStampedModel):
+            entitlement = cls.objects.get(
+                user=user,
+                course_uuid=course_uuid,
+            )
```
Can a user have more than one entitlement for a course? (Maybe both 'credit' and 'verified'?

staubina replied:
This all relies on the assumption a course will only ever have one paid mode.  Also Credit is not a great example because of how that is applied. 

This question makes sense when thinking about 'verified' and 'professional' though.

I will double check, but I believe a Course Entitlement and its runs all should have the same modes.

staubina replied:
This method is dead code. Deleting.

amangano-edx commented on common/djangoapps/entitlements/models.py:
```diff
@@ -24,3 +24,25 @@ class CourseEntitlement(TimeStampedModel):
     order_number = models.CharField(max_length=128, null=True)
+
+    @classmethod
+    def get_active_user_course_entitlements(cls, user, course_uuid):
```
Nit: The method name is plural, but the logic only seems like it would ever return a single entitlement.

staubina replied:
It will return multiple entitlements.

amangano-edx commented on common/djangoapps/entitlements/models.py:
```diff
@@ -24,3 +24,25 @@ class CourseEntitlement(TimeStampedModel):
+        """
+        Fulfills an entitlement by specifying a session.
+        """
+        cls.objects.filter(id=entitlement.id).update(enrollment_course_run=enrollment)
```
Should we do any validation here to make sure the entitlement hasn't already been expired?

staubina replied:
I think that check belongs at the higher levels that would be making this call.  Potentially before the enroll happens.  We will add this with integration of the policy management code.

staubina replied:
I am adding AC to the integration task to make sure we add in checks on enroll and any we think we need on switch session.

staubina replied:
Thinking about this some more.  I will let you know when I push up changes.

amangano-edx commented on common/djangoapps/entitlements/migrations/0003_auto_20171120_1432.py:
```diff
@@ -0,0 +1,20 @@
+        migrations.AlterField(
+            model_name='courseentitlement',
+            name='uuid',
+            field=models.UUIDField(default=uuid.uuid4, unique=True, editable=False),
```
Is there any chance that this could cause issues, if there is data in the database already that isn't unique?

staubina replied:
There is nothing the Database currently. We do not have entitlements in prod.

staubina replied:
Also this needs to be updated once Jeff's PR merges to masted.

amangano-edx replied:
I've think there might be data on stage, so we may want to double check that it is valid before this runs.

jlajoie replied:
Will this have negative impacts on open source at all? Just because we don't have data in Prod and can drop data in stage at will, doesn't mean another group might have that option. Unless this feature isn't "released" yet and no one's supposed to be using it?

staubina replied:
I think if this was a field in the database for a long period of time I would be more concerned, but no one is using it currently.  Unless someone has manually entered a field in the Database on Stage then we should not have a non unique entry. This is why we need to make this change now.  I will take a look at the read replica on Stage before merging.

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+    authentication_classes = (JwtAuthentication, SessionAuthentication,)
+    permission_classes = (permissions.IsAuthenticated,)
+    queryset = CourseEntitlement.objects.all()
+    serializer_class = CourseEntitlementSerializer
```
Since this is an enrollment endpoint, would it make sense to also serialize the enrollment that gets created/updated/destroyed?

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+            user=user,
+            course_key=course_session_key,
+            mode=entitlement.mode,
+        )
```
Does this save to the DB immediately? If so, it should probably be run in a transaction along with https://github.com/edx/edx-platform/pull/16617/files#diff-9b73749954d13e21a49b7ea1e0b110a9R72

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+
+    def _unenroll_entitlement(self, entitlement, course_session_key, user):
+        CourseEnrollment.unenroll(user, course_session_key, skip_refund=True)
+        CourseEntitlement.set_enrollment(entitlement, None)
```
These calls should probably be run in a transaction if they're modifying DB entries.

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+        CourseEnrollment.unenroll(user, course_session_key, skip_refund=True)
+        CourseEntitlement.set_enrollment(entitlement, None)
+
+    def create(self, request, uuid):
```
Nit: it would be helpful to have a docstring here.

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+        if not course_session_id:
+            return Response(
+                status=status.HTTP_400_BAD_REQUEST,
+                data="The Course Run ID was not provided."
```
Should this text be translated?

amangano-edx replied:
Note: when translating these we may need to escape the translations differently since they'll be provided via the JSON? response.

staubina replied:
We do not need to translate strings meant for Devs in the API.  This is not meant for display.

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+        except CourseEntitlement.DoesNotExist:
+            return Response(
+                status=status.HTTP_400_BAD_REQUEST,
+                data="The Entitlement for this UUID does not exist or is Expired."
```
Should this be translated?

staubina replied:
Same as above.

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+        course_runs = get_course_runs_for_course(entitlement.course_uuid)
+        for run in course_runs:
+            if course_session_id == run.get('key', ''):
+                course_run_valid = True
```
Do all of the course_runs need to match, or just one? If it's just one you should be able to break out of the loop early as soon as you find one. You also might be able to shorten this by using the `all` method.
`all(lambda run: course_session_id == run.get('key', ''), course_runs)`

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+        if not course_run_valid:
+            return Response(
+                status=status.HTTP_400_BAD_REQUEST,
+                data="The Course Run ID is not a match for this Course Entitlement."
```
Should we translate this?

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+                user=request.user
+            )
+        else:
+            if entitlement.enrollment_course_run.course_id != course_session_id:
```
Nit: this could be moved to line 116 and changed to `elif`

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+                    entitlement=entitlement,
+                    course_session_key=CourseKey.from_string(course_session_id),
+                    user=request.user
+                )
```
This calls should probably be made within a transaction.

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+        course_run_valid = False
+        course_runs = get_course_runs_for_course(entitlement.course_uuid)
+        for run in course_runs:
+            if course_session_id == run.get('key', ''):
```
Do we need to normalize the types before we make this comparison? (`str` vs `unicode`)?

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+                )
+
+        return Response(
+            status=status.HTTP_201_CREATED,
```
Would it make sense to return a different status for update vs create so that the client knows what happened on the backend? Maybe 200 for the update case and 201 for create? 

Also, if the calls to _enroll/_unenroll fail for some reason, what will be returned?

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+            return Response(
+                status=status.HTTP_400_BAD_REQUEST,
+                data="The Entitlement for this UUID does not exist or is Expired."
+            )
```
Nit: This block is repeated in the `create` method. Could it be DRYd up somehow?

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -51,3 +54,105 @@ def perform_destroy(self, instance):
+        except CourseEntitlement.DoesNotExist:
+            return Response(
+                status=status.HTTP_400_BAD_REQUEST,
+                data="The Entitlement for this UUID does not exist or is Expired."
```
This should probably be translated.

staubina replied:
Same as above.

McKenzieW commented:
jenkins run python

McKenzieW commented:
jenkins run quality

mjfrey commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -57,3 +64,167 @@ def perform_destroy(self, instance):
+        Internal method to handle the details of Unenrolling a User in a Course Run.
+        """
+        CourseEnrollment.unenroll(user, course_run_key, skip_refund=True)
+        CourseEntitlement.set_enrollment(entitlement, None)
```
Shouldn't unenroll take care of this itself?

staubina replied:
I chose to not mess with the CourseEnrollment code, as it is specific to Course Runs.  This separates the two. I can change this if we are concerned.

jlajoie commented on common/djangoapps/entitlements/api/v1/tests/test_views.py:
```diff
@@ -195,3 +198,168 @@ def test_revoke_unenroll_entitlement(self):
+            data=json.dumps(data),
+            content_type='application/json',
+        )
+        course_entitlement.refresh_from_db()  # pylint: disable=no-member
```
I don't think you need to disable the warning here if you swap `CourseEntitlementFactory()` to use `CourseEntitlementFactory.create()`

jlajoie commented on common/djangoapps/entitlements/api/v1/tests/test_views.py:
```diff
@@ -195,3 +198,168 @@ def test_revoke_unenroll_entitlement(self):
+        super(EntitlementEnrollmentViewSetTest, self).tearDown()
+
+    def test_user_can_enroll(self):
+        course_entitlement = CourseEntitlementFactory(
```
Tying in with below, can probably use the .create() call here to avoid the need for disabling linters and making this consistent with the other tests

jlajoie commented on common/djangoapps/entitlements/api/v1/tests/test_views.py:
```diff
@@ -195,3 +198,168 @@ def test_revoke_unenroll_entitlement(self):
+        )
+        course_entitlement.refresh_from_db()
+
+        assert response.status_code == 201
```
Just checking that this is intended, do we use 201 here for updates as well as creates?

staubina replied:
Yes, that is the intention. 201 always for a successful Enroll.

jlajoie commented on common/djangoapps/entitlements/api/v1/urls.py:
```diff
@@ -1,11 +1,22 @@
 urlpatterns = [
     url(r'', include(router.urls)),
+    url(
+        r'entitlements/(?P<uuid>[0-9a-f-]+)/enrollments/$',
```
This regex doesn't reflect an actual UUID, you can just as easily have something like `entitlements/---/enrollments/ evaluate to true in here. 

Separate thought, is it worth enforcing the trailing slash?

jlajoie commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -57,3 +64,167 @@ def perform_destroy(self, instance):
+        for run in course_runs:
+            if course_run_id == run.get('key', ''):
+                course_run_valid = True
+                break
```
Nit: Personally I hate breaks and think they're icky, since we're not doing anything with the course_run_id we could swap this to 

```
for run in course_runs:
    if course_run_id == run.get('key', ''):
        return True
return False
```

staubina commented on common/djangoapps/entitlements/migrations/0003_auto_20171120_1432.py:
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
```
Note: This needs to be removed before merging.  It could clash with another migration.  After the Policy manager PR is merged I will rebase on master and create a new migration.

staubina replied:
Not removed, but recreated.

jlajoie commented:
Just those few comments, otherwise looks reasonable

staubina commented:
@jlajoie and @mjfrey  Please take another look when you have a moment.  I think I got all your requests.

amangano-edx commented on common/djangoapps/entitlements/api/v1/urls.py:
```diff
@@ -1,11 +1,22 @@
 urlpatterns = [
     url(r'', include(router.urls)),
+    url(
+        r'entitlements/(?P<uuid>[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12})/enrollments$',
```
It might be nice to save this regex to a constant so that we can re-use it elsewhere, much like we do for the COURSE_ID_PATTERN.

bderusha replied:
Agreed.  Can we pull this out into a constant.  Having a reusable UUID would be really helpful

amangano-edx commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -3,22 +3,29 @@
     permission_classes = (permissions.IsAuthenticated, IsAdminOrAuthenticatedReadOnly,)
-    lookup_value_regex = '[0-9a-f-]+'
+    queryset = CourseEntitlement.objects.all().select_related('user')
+    lookup_value_regex = '[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}'
```
This is repeated in urls.py. It would be nice if it were stored in a constant somewhere so that we can keep the code DRY.

staubina commented:
jenkins run bokchoy

bderusha commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -57,3 +64,165 @@ def perform_destroy(self, instance):
+
+        return Response(
+            status=status.HTTP_201_CREATED,
+            data={
```
I'm fine with this.  But can we put a comment above noting that the serializer is not used.

Can we get away with not setting a serializer if we don't use it?

bderusha commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -57,3 +64,165 @@ def perform_destroy(self, instance):
+        return Response(
+            status=status.HTTP_201_CREATED,
+            data={
+                'uuid': entitlement.uuid,
```
should be `entitlement_uuid` as this is technically an "enrollment" object

staubina commented:
@bderusha Please take a look, I believe I addressed all your concerns.  I have also corrected the Python test failures.

FYI: @mjfrey @jlajoie There are some changes in the returned Responses, but nothing significant.  Please let me know if you have concerns.

staubina commented:
I also updated the migration to remove the naming conflict.

bderusha commented on common/djangoapps/entitlements/api/v1/urls.py:
```diff
@@ -1,11 +1,22 @@
 urlpatterns = [
     url(r'', include(router.urls)),
+    url(
+        r'entitlements/(?P<uuid>{regex})/enrollments$'.format(regex=EntitlementViewSet.ENTITLEMENT_UUID4_REGEX),
```
🎸

bderusha commented on common/djangoapps/entitlements/api/v1/views.py:
```diff
@@ -102,3 +109,171 @@ def perform_destroy(self, instance):
+
+        return Response(
+            status=status.HTTP_201_CREATED,
+            data=data
```
This is unexpected to me.  This endpoint is dealing with enrollments not entitlements.  I don't think that we should be returning an entitlement here.

I think the way you had it before was fin, returning en enrollment object with `entitlement_uuid`, `course_run_id`, and `active` flag. 

The true right way is to create an `EntitlementEnrollmentSerializer` and wire it up like this.  For expediency I think returning the hand crafted dict is fine too.  Given that all this exists it may be easier to "fix forward" and just create the serializer anyway

staubina replied:
I can return it to that.  We can discuss this in more detail tomorrow.

staubina replied:
Also, I do not need to send the is_active flag anymore. That was a remnant of our initial implementation.
